bin/pg_repack.c

```diff
@@ -2073,14 +2113,47 @@ repack_all_indexes(char *errbuf, size_t errsize)
 
 		cell = r_index.head;
 	}
-	else if (table_list.head)
+	else if (table_list.head || parent_table_list.head)
 	{
 		appendStringInfoString(&sql,
 			"SELECT i.relname, idx.indexrelid, idx.indisvalid, idx.indrelid, $1::text, n.nspname"
 			" FROM pg_index idx JOIN pg_class i ON i.oid = idx.indexrelid"
 			" JOIN pg_namespace n ON n.oid = i.relnamespace"
 			" WHERE idx.indrelid = $1::regclass ORDER BY indisvalid DESC, i.relname, n.nspname");
 
+		for (cell = parent_table_list.head; cell; cell = cell->next)
+		{
+			int nchildren, i;
+
+			params[0] = cell->val;
+
+			/* find children of this parent table */
+			res = execute_elevel("SELECT quote_ident(n.nspname) || '.' || quote_ident(c.relname)"
+								 " FROM pg_class c JOIN pg_namespace n on n.oid = c.relnamespace"
+								 " WHERE c.oid = ANY (repack.get_table_and_inheritors($1::regclass))"
+								 " ORDER BY n.nspname, c.relname", 1, params, DEBUG2);
+
+			if (PQresultStatus(res) != PGRES_TUPLES_OK)
+			{
+				elog(WARNING, "%s", PQerrorMessage(connection));
+				continue;
+			}
+
+			nchildren = PQntuples(res);
+
+			if (nchildren == 0)
```

Comment:
@MasahikoSawada,

Totally forgot to describe possible changes to `repack_get_table_and_inheritors()`.

We could, for example, take a lock on parent and traverse children tree with `AccessShareLock`:

```C
Datum
repack_get_table_and_inheritors(PG_FUNCTION_ARGS)
{
	Oid	    parent = PG_GETARG_OID(0);
	List	   *relations;
	Datum	   *relations_array;
	int	    relations_array_size;
	ArrayType  *result;
	ListCell   *lc;
	int         i;

        /* take locks HERE! */
        LockRelationOid(parent, AccessShareLock);
	relations = find_all_inheritors(parent, AccessShareLock, NULL);

	relations_array_size = list_length(relations);
	if (relations_array_size == 0)
		PG_RETURN_ARRAYTYPE_P(construct_empty_array(OIDOID));
```

But the problem remains: the `parent` table might be dead when we finally get to the ` LockRelationOid(parent, AccessShareLock);` since `'TABLE_NAME'::regclass` won't lock it.